vortex-array/src/array/primitive/mod.rs

```diff
@@ -191,3 +194,43 @@ impl<'a> Array<'a> {
 }
 
 impl EncodingCompression for PrimitiveEncoding {}
+
+impl ScalarSubtractFn for PrimitiveArray<'_> {
+    fn scalar_subtract(&self, to_subtract: Scalar) -> VortexResult<OwnedArray> {
+        if self.dtype() != to_subtract.dtype() {
+            vortex_bail!(MismatchedTypes: self.dtype(), to_subtract.dtype())
+        }
+
+        let result = match to_subtract.dtype() {
+            DType::Int(..) => {
+                match_each_integer_ptype!(self.ptype(), |$T| {
+                    let to_subtract = <scalar::Scalar as TryInto<$T>>::try_into(to_subtract)?;
+                    let maybe_min = self.statistics().compute_as_cast(Stat::Min);//.unwrap_or($T::MAX);
+
+                    if maybe_min.is_some() {
+                        let min: $T = maybe_min.unwrap();
+                        let max: $T = self.statistics().compute_as_cast(Stat::Max).unwrap();
+                        if let (min, true) = min.overflowing_sub(to_subtract) {
+                            vortex_bail!("Integer subtraction over/underflow: {}, {}", min, to_subtract)
+                        }
+                        if let (max, true) = max.overflowing_sub(to_subtract) {
+                                vortex_bail!("Integer subtraction over/underflow: {}, {}", max, to_subtract)
+                        }
+                    }
+                    let sub_vec : Vec<$T> = self.typed_data::<$T>().iter().map(|&v| v - to_subtract).collect_vec();
+                    PrimitiveArray::from(sub_vec)
+                })
+            }
+            DType::Decimal(..) | DType::Float(..) => {
+                match_each_native_ptype!(self.ptype(), |$T| {
```

Comment:
Yeah on second thought it's probably worth adding a match_each_float_type. I figured we could assume this would only produce floats (otherwise we have a bug), but it's probably worth adding a check to panick if that's not true.

Reply:
More specifically, the body of this macro currently _must_ compile for all PTypes. So you cannot write any code that is specific to floats. It just so happens that the subtract operator works for all these.